logger_test.go

```diff
@@ -385,6 +385,57 @@ func TestLoggerAddCallerFail(t *testing.T) {
 	})
 }
 
+func TestLoggerAddFunction(t *testing.T) {
+	tests := []struct {
+		options []Option
+		pat     string
+	}{
+		{opts(), `^$`},
+		{opts(WithFunction(false)), `^$`},
+		{opts(AddFunction()), `zap.TestLoggerAddFunction.func1$`},
+		{opts(AddFunction(), WithFunction(false)), `^$`},
+		{opts(WithFunction(true)), `zap.TestLoggerAddFunction.func1$`},
+		{opts(WithFunction(true), WithFunction(false)), `^$`},
+		{opts(AddFunction(), AddCallerSkip(1), AddCallerSkip(-1)), `zap.TestLoggerAddFunction.func1$`},
+		{opts(AddFunction(), AddCallerSkip(1)), `zap.withLogger$`},
+		{opts(AddFunction(), AddCallerSkip(1), AddCallerSkip(3)), `runtime.goexit$`},
+	}
+	for _, tt := range tests {
+		withLogger(t, DebugLevel, tt.options, func(logger *Logger, logs *observer.ObservedLogs) {
+			// Make sure that sugaring and desugaring resets caller skip properly.
+			logger = logger.Sugar().Desugar()
+			logger.Info("")
```

Comment:
I've updated the code to include `logger.Sugar().Info("")` and `logger.Sugar().Desugar().Info("")` in caller function test. However, in the tests table I ended up defining `loggerFunction` and `sugaredFunction` which seems to be redundant, but I defined them because I need to use different log function for normal and sugared logger.